api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			return
+		}
+		c, err := cid.Decode(lastestMsgCid.Cid)
+		if err != nil {
+			tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+			tx.FailureMsg = fmt.Sprintf("decoding latest message cid for txn: %v", err)
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			return
+		}
+
+		tx.Waiting = true
+		if err := s.updateTxn(ctx, tx); err != nil {
+			ch <- waitResult{err: err}
+			return
+		}
+
+		res, err := client.StateWaitMsg(ctx, c, s.config.MessageConfidence)
+		tx.Waiting = false
+		if err != nil {
+			// If for some reason the lotus node doesn't know about the cid, consider that a final error.
+			if strings.Contains(err.Error(), "block not found") {
+				tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+				tx.FailureMsg = err.Error()
+				tx.UpdatedAt = time.Now()
+				log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			}
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			if tx.MessageState == pb.MessageState_MESSAGE_STATE_FAILED {
+				ch <- waitResult{txn: tx}
+			} else {
+				ch <- waitResult{err: fmt.Errorf("calling StateWaitMsg: %v", err)}
+			}
+			return
+		}
+
+		if res.Receipt.ExitCode.IsError() {
+			tx.MessageState = pb.MessageState_MESSAGE_STATE_FAILED
+			tx.FailureMsg = fmt.Sprintf("error exit code: %v", res.Receipt.ExitCode.Error())
+			tx.UpdatedAt = time.Now()
+			log.Warnf("failing txn with err: %s", tx.FailureMsg)
+			if err := s.updateTxn(ctx, tx); err != nil {
+				ch <- waitResult{err: err}
+				return
+			}
+			ch <- waitResult{txn: tx}
+			if res.Receipt.ExitCode.IsSendFailure() {
+				log.Errorf("received exit code send failure: %s", res.Receipt.ExitCode.String())
+			} else {
+				log.Infof("received exit code error: %s", res.Receipt.ExitCode.String())
+			}
+			return
+		}
+
+		tx.MessageState = pb.MessageState_MESSAGE_STATE_ACTIVE
+		tx.UpdatedAt = time.Now()
+
+		// This would probably not ever be true because we would already know about and have tracked
+		// the new message cid if we replaced the message with a new one. Checking just in case.
+		isNewCid := true
+		for _, msgCid := range tx.MessageCids {
+			if res.Message.String() == msgCid.Cid {
+				isNewCid = false
+				break
+			}
+		}
+
+		if isNewCid {
+			tx.MessageCids = append(tx.MessageCids, msgCid{Cid: res.Message.String(), CreatedAt: time.Now()})
+		}
+
+		if err := s.updateTxn(ctx, tx); err != nil {
+			ch <- waitResult{err: err}
+			return
+		}
+
+		s.waitingLck.Lock()
+		delete(s.waiting, tx.ID)
+		s.waitingLck.Unlock()
+
+		ch <- waitResult{txn: tx}
```

Comment:
Remove the chan from the in memory map and send the final txn on the results chan.